src/progress.rs

```diff
@@ -197,15 +195,13 @@ impl ProgressSet {
 
     /// Promote a learner to a peer.
     pub fn promote_learner(&mut self, id: u64) -> Result<(), Error> {
-        match self.progress.get_mut(&id) {
-            Some(progress) => if !progress.is_learner {
-                Err(Error::Exists(id, "voters"))?;
-            } else {
-                progress.is_learner = false;
-                self.configuration.voters.insert(id);
-                self.configuration.learners.remove(&id);
-            },
-            None => Err(Error::NotExists(id, "learners"))?,
+        if self.learner_ids().contains(&id) {
```

Comment:
Why not just `remove`?

Reply:
Yes, I could check if it is `Some(v)` and put it back, returning an error. Good idea. :)